### ISSUE TITLE
Network connectivity issues with iOS devices 14.2.0+ & ASA SDK 2.7.1

### ISSUE THREAD

**Issue by git-radiusmedia** [Azure/azure-spatial-anchors-samples#234]
With azure spatial anchors sample basic demo, the iOS devices running 14.2 or higher have a problem finding the anchor after some attempts(varying). After successfully finding the anchor some times, suddenly the internet connectivity to the app is lost. I am still not sure if it is the problem with the iOS update or ASA SDK.

We noticed it when trying to get info in the app from Databases and suddenly it stops working and doesn't work anymore. 

This I have noticed in iPad Pro, iOS 14.2, Unity 2019.4.17f1, ASA SDK 2.7.1 & 2.7.0, Azure services region is Europe.@msftradford

**Comment by msftradford** [comment]
Thanks for filing @git-radiusmedia! About how many times can you successfully restart the basic demo mode before you have an issue? Have you noticed the same issue on iOS version <14.2? Have you been able to test on other platforms, like Android? Have you tested using the native versions (non-Unity) of the sample app? Were you having similar issues before upgrading to ASA 2.7.0?  Can you reproduce the issue by disabling your networking while using the app? Does the issue happen with just the basic demo mode or does it happen in other modes as well? Would you mind sending me a log from a session that resulted in this issue?

@ephb, are you experiencing similar issues?

**Comment by ian-phoria** [comment]
Just chiming in to say we've been having the exact same issue since late last year with versions prior to ASA 2.7.0. Running in Unity 2020.1.16f1. The issue is exactly the same as above. Resolving works for multiple attempts and then suddenly stops working - from that point on, the app loses all internet connectivity until you restart. This is in our own application (not the sample app) that was working flawlessly prior to some point in November/December. 

It only occurs for us on iOS devices running iOS 14.2. We tested on both Android devices and iOS devices running versions prior to iOS 14 and they both worked as expected. 

Azure services region for us is Australia East. 

I have just updated to ASA SDK 2.7.1 so will test and report back with clearer information (and logs) if the issue persists.

**Comment by msftradford** [comment]
Thanks for reporting-in @ian-phoria! We look forward to the results of your experiments. You mentioned that the issue happens in your own app, but have you tried to reproduce it using any of the ASA sample apps on a device that is known to have this issue?

**Comment by git-radiusmedia** [comment]
Hello,
Just to be clear we are not using the basic demo samples but it's in our own application that uses the concept just like in the basic demo but hasn't tested with the sample project. I have tested on some of the iOS devices with version <14.2 and it worked on all the devices without any problem(even though some of the devices are really low performing) and also on Android devices with no problem at all. Never tried the native version as I work with Unity. The same problem arose even when working with ASA SDK 2.5.0 and iOS devices with version >14.2. As I have mentioned above, our app needs to be connected to the database to get information so haven't tried working without internet connectivity.

**Comment by ephb** [comment]
This sounds exactly like my issue but we have not done any extended testing yet. 

This behavior seems to be in line with what @cmelian reported in the other [issue](https://github.com/Azure/azure-spatial-anchors-samples/issues/228#issuecomment-742345762). So I think one could take the same steps he did to check if the servers certificates are still the cause.

To be clear we are also not using the sample but our own application, on Azure services region Europe.

It is noteworthy that this will prevent all communication of the app with the internet after the error showed up. Not just to ASA, our bug tracker and database connection will also fail. Connecting to the devices console will show ssl errors after this occurred as far as I can remember.

One problem is that the change notes of the SDK do not go into a lot of detail. At least from the mentioned changes I would not assume the issue to be fixed.

**Comment by msftradford** [comment]
Thank you all very much for the feedback.

@ian-phoria and @ephb, we look forward to hearing the result of your experiments with ASA SDK 2.7.1.

@git-radiusmedia, you mentioned that the “internet connectivity to the app was lost”, do you know at which stage of the ASA pipeline the error occurs? For example, does happen it while scanning the environment, uploading an anchor, locating an anchor etc. Have you been able to take a look at the logs? When you mention that the connection to a database stops working, can you explain the type of connection or what networking library is used?

We would like to explain a bit more about the issues that were addressed in ASA SDK 2.7.1. We noted that, beginning with iOS 14.2.0, the mechanism used for running concurrent code on iOS, [Dispatch using GCD queues](https://developer.apple.com/documentation/DISPATCH), would stop servicing the ASA SDK events being placed into the queue. This would manifest as [progress values](https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.spatialanchors.sessionstatus.readyforcreateprogress?view=spatialanchors-dotnet) passed back by the SDK in the [SessionStatus](https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.spatialanchors.sessionstatus?view=spatialanchors-dotnet) class to not increment even though user was properly moving the device and scanning environment, or save or delete anchors SDK callbacks never being raised to app code even though anchors were successfully saved/deleted in the cloud. If other libraries used by the app (e.g., networking libs) use these GCD queues, they would most likely stop working as well. 

We mitigated this issue by limiting the number of events sent to the queue and by changing the queue identifier passed to the [dispatch_async()](https://developer.apple.com/documentation/dispatch/1453057-dispatch_async) function in SDK 2.7.1 release. Prior to this release, ASA SDK used the [DISPATCH_QUEUE_PRIORITY_DEFAULT](https://developer.apple.com/documentation/dispatch/dispatch_queue_priority_default) queue identifier and in 2.7.1 it was changed to use the [QOS_CLASS_UTILITY ](https://developer.apple.com/documentation/dispatch/dispatch_queue_priority_low) identifier. 

If you are experiencing similar issues with your ASA SDK-powered app, please let us know if you have noticed whether the GCD queues have stopped being serviced. Once the app is in a bad state, you can pause execution in XCode and take a look at [“View Process by Queue”](https://developer.apple.com/library/archive/documentation/DeveloperTools/Conceptual/debugging_with_xcode/chapters/debugging_tools.html) in “Process View Display” in the debugger and check to see if there are too many blocks waiting for execution. It is worth checking if your networking library is using GCD on iOS devices. Also, it would be very helpful to test if the same network connectivity issue happens if the ASA SDK is disabled in order to pinpoint that the issue has a root cause in the ASA SDK. 

We have opened up an internal bug (#31330257) to look into this issue further.

**Comment by ephb** [comment]
Thank you very much @msftradford for the detailed explanation. 

I will report back when I have looked into it more.
Is there any way I can share a log privately with you or the team looking into this?

**Comment by msftradford** [comment]
Yes, that would be quite helpful. I'll send you a direct email momentarily.

**Comment by git-radiusmedia** [comment]
Hello,
This occurred during looking for an anchor. I am using Unity networking to connect to my Database.

**Comment by msftradford** [comment]
Thanks very much for the responses @ephb, @ian-phoria, and @git-radiusmedia. We added a networking component to our sample apps and were able to reproduce what we believe to be this issue. Please be aware that consumers of the [ARFoundation samples](https://github.com/Unity-Technologies/arfoundation-samples) are [experiencing similar issues](https://github.com/Unity-Technologies/arfoundation-samples/issues/694). We took our same networking component and applied it to the ARFoundation samples, without using the ASA SDK, and were able to reproduce the same issue.

We believe that this issue is related to another [reported issue](https://github.com/Azure/azure-spatial-anchors-samples/issues/228) between ASA SDK and iOS 14.2+. Taking the same solution that we used to address that issue in our SDK and applying it to the generated Unity code in the XCode project has been suggested as a workaround. 

To those customers that are experiencing this issue, please open your generated XCode project and find the native function called `UnityCreateWebRequestBackend()`. After the following lines:

```
webOperationQueue = [[NSOperationQueue alloc] init];
webOperationQueue.name = @"com.unity3d.WebOperationQueue";
```

Add the line:

```
webOperationQueue.qualityOfService = NSQualityOfServiceUtility;
```

Please let us know how things go after making the change. At the same time we will work to find a more permanent solution. Thanks!

**Comment by ephb** [comment]
Hi @msftradford,
thank you for the investigation and detailed description.
I will try this tomorrow. I hadn’t emailed you yet as didn’t have much luck reproducing the error currently with 2.7.1.
If these changes produce noticeable change it should be possible to create a post-build script to automate these steps for now.

**Comment by ian-phoria** [comment]
This solution seems to have fixed the issue on my side - though I did have similar issues with reproducing my initial error on ASA SDK 2.7.1.

Just the upgrade to ASA SDK 2.7.1 seemed to have solved a bunch of my "failed to find anchor when I'm right in front of it" issues (went from max 4 successful locates before the string of failures started to 15+ locates in a row) - but the internet connectivity loss still occurred on iOS 14.2.

Adding the extra line to UnityCreateWebRequestBackend() has solved the problem completely. Just need to set up a post-build script so we can run this on Unity Cloud build. Thank you all (on both sides of the discussion) for the investigation and solution.

**Comment by ephb** [comment]
Hi @msftradford,
we finally got to looking into this again and would occasionally get the connection problems unless we used the workaround you suggested.
Did you submit this problem to Unity yet? Otherwise I create a bug report.

**Comment by msftradford** [comment]
Thanks for checking-in @ephb! Unity is aware of the issue but we do not know much more than what is mentioned [here](https://github.com/Unity-Technologies/arfoundation-samples/issues/694).

**Comment by devSebas** [comment]
Hello,
This issue has been official fixed on 2020.2.4 and 2019.4.21 LTS.
Please close the issue.
